### DISCUSSION TITLE
Is the incremental APK generated by the android_binary rule supposed to include the resources in the primary APK?

### DISCUSSION THREAD

**Discussion by Hounshell** [bazelbuild/bazel#18750] [Q&A]
I'm looking at the performance of Android builds, particularly for large apps that are being developed on cloud developer instances. I though that incremental installs via `bazel mobile-install` would be killer in this environment, but we're not seeing improvements at all. We created a simple app with a single very large resource (foo.bin) to test this out and look at timing.

Reading between the lines in incremental_install.py lines 775 - 784, it looks like the steps are basically:

- Copy the dexes to the device
- Copy resources to the device
- Copy native libraries to the device
- Copy the _incremental.apk to the device

Tweaking things and running them generates expected output (updating code indicates that a dex is being copied, but resources are not, updating resources indicates that resources are being copied but a dex is not). But that last step is where we're hitting a snag. The _incremental.apk contains all of the resources. Looking at the contents of that file I see:

```
 Length   Method    Size  Cmpr    Date    Time   CRC-32   Name
--------  ------  ------- ---- ---------- ----- --------  ----
      57  Defl:N       54   5% 01-01-2010 00:00 02fdd058  stub_application_data.txt
   18632  Defl:N     9222  51% 01-01-2010 00:00 d4a25f11  classes.dex
    3324  Defl:N     1073  68% 01-01-2010 00:00 939ef210  AndroidManifest.xml
    1873  Stored     1873   0% 01-01-2010 00:00 a72b917b  res/drawable-mdpi-v4/ic_launcher.png
    3090  Stored     3090   0% 01-01-2010 00:00 afa02fad  res/drawable-hdpi-v4/ic_launcher.png
    4289  Stored     4289   0% 01-01-2010 00:00 23985095  res/drawable-xhdpi-v4/ic_launcher.png
    7423  Stored     7423   0% 01-01-2010 00:00 e1a7949e  res/drawable-xxhdpi-v4/ic_launcher.png
     227  Stored      227   0% 01-01-2010 00:00 81d59299  res/drawable-hdpi-v4/tile.9.png
     956  Defl:N      444  54% 01-01-2010 00:00 ffa6e782  res/layout/sample_main.xml
2000000000  Stored 2000000000   0% 01-01-2010 00:00 3d09ea7a  res/raw/foo.bin
    3772  Stored     3772   0% 01-01-2010 00:00 07cf7063  resources.arsc
    1004  Defl:N      527  48% 01-01-2010 00:00 5efc6d0a  META-INF/CERT.SF
    1203  Defl:N     1052  13% 01-01-2010 00:00 bff25f97  META-INF/CERT.RSA
     915  Defl:N      473  48% 01-01-2010 00:00 7fc4c009  META-INF/MANIFEST.MF
--------          -------  ---                            -------
2000046765         2000033519   0%                            14 files
```

So clearly my foo.bin is bloating the size of this incremental apk. 

Is this intentional? Is there a way to ensure that this apk has only the bare minimum resources necessary so that we don't have to transfer this on every install?

**Comment by andrejsuslov749** [comment]
interesting!
